### DISCUSSION TITLE
How to configure unique IP addresses for WSL2 instances?

### DISCUSSION THREAD

**Discussion by bimmerdriver** [microsoft/WSL#12069] [Q&A]
I'm running WSL2 on Windows 10 22H2. I have an existing instance of Ubuntu 22.04 and I want to migrate it to Ubuntu 24.04. I want to leave the 22.04 instance alone while I'm setting up the new 24.04 instance. I created the 24.04 instance and I noticed that it has the same IP address as the 22.04 instance. Why is this? It makes no sense that both instances should have the same IP address when they are clearly distinct from each other. Is there a way to change the IP address of the 24.04 instance?

**Comment by bimmerdriver** [comment]
Anyone?

**Comment by reini-1** [comment]
I stumbled over the same problem.
Did you find a solution for this?

**Comment by brasswood** [comment]
Answered by #4304

**Reply by bimmerdriver** [reply]
Thanks for the link, but it's disappointing to find out that this behavior is by design. It makes no sense for instances to share the same IP address.

**Reply by mattcheau** [reply]
@bimmerdriver That's not really what that link says at all.

**Reply by bimmerdriver** [reply]
@mattcheau I just read the link from start to finish. English is my first language and I think I have pretty good comprehension. It seems pretty clear that it's by design that WSL2 instances share the same IP address by design. I would like to access web servers hosted by each WSL instance using unique, dedicated IP addresses and port numbers. I'm not the only one who is asking for this to change.

**Reply by pxn** [reply]
@bimmerdriver and that is exactly reason - multiple webservers and no license money for IIS / whatever MS else

**Comment by junaiddeveloper786** [comment]
@bimmerdriver 


🌐 Why WSL2 Instances Share the Same IP (and How to Configure Unique Ones)
⚙️ Why This Happens

WSL2 doesn’t run VMs like VirtualBox/VMware with individual NICs.

Instead, it uses a single lightweight utility VM with a virtual switch (Hyper-V) that hosts all your Linux distributions.

Each distro (Ubuntu 22.04, 24.04, etc.) is isolated at the process level, but they share the same network namespace, so they appear under the same internal IP.

That’s why your Ubuntu 22.04 and 24.04 both show the same IP — they’re riding on the same network interface provided by the WSL2 VM.

🛠️ Options to Assign Unique IPs

Use localhost/port binding (Recommended for most use cases)

Instead of relying on per-distro IPs, bind each service to a unique port.

Example:

Ubuntu 22.04 service → localhost:8080

Ubuntu 24.04 service → localhost:9090

WSL automatically handles port forwarding from Windows to the right instance.

Enable WSL2 networking with separate NAT adapters (Experimental / Hacky)

You can configure separate Hyper-V NAT networks and attach different WSL distros to them using wsl.conf and custom scripts.

Example: Create a new vSwitch and assign it to a distro manually.

⚠️ This is non-trivial and not officially supported.

Run one instance under WSL2 and another under WSL1

WSL1 instances have their own unique networking model (they use Windows’ network stack directly).

This way you can run 22.04 on WSL2 and 24.04 on WSL1 with distinct IP behavior.

Use Docker or VM approach for strict IP separation

If you truly need per-distro unique IPs (like for complex multi-host setups), it’s cleaner to use Docker networks or full Hyper-V VMs.

📌 Bottom Line

By design, WSL2 distros share the same virtualized network interface → same IP.

For most workflows, unique ports per instance is the recommended solution.

If you absolutely need unique IPs, you’ll have to rely on advanced networking (custom Hyper-V NAT) or switch one distro to WSL1.

✅ In short: WSL2 isn’t designed to give each distro its own IP — they all share the same VM’s network namespace. The practical solution is to differentiate services by ports or use WSL1/VMs for strict per-distro IP isolation.

**Reply by bluebrown** [reply]
> @bimmerdriver
> 
> 🌐 Why WSL2 Instances Share the Same IP (and How to Configure Unique Ones) ⚙️ Why This Happens
> 
> WSL2 doesn’t run VMs like VirtualBox/VMware with individual NICs.
> 
> Instead, it uses a single lightweight utility VM with a virtual switch (Hyper-V) that hosts all your Linux distributions.
> 
> Each distro (Ubuntu 22.04, 24.04, etc.) is isolated at the process level, but they share the same network namespace, so they appear under the same internal IP.
> 
> That’s why your Ubuntu 22.04 and 24.04 both show the same IP — they’re riding on the same network interface provided by the WSL2 VM.
> 
> 🛠️ Options to Assign Unique IPs
> 
> Use localhost/port binding (Recommended for most use cases)
> 
> Instead of relying on per-distro IPs, bind each service to a unique port.
> 
> Example:
> 
> Ubuntu 22.04 service → localhost:8080
> 
> Ubuntu 24.04 service → localhost:9090
> 
> WSL automatically handles port forwarding from Windows to the right instance.
> 
> Enable WSL2 networking with separate NAT adapters (Experimental / Hacky)
> 
> You can configure separate Hyper-V NAT networks and attach different WSL distros to them using wsl.conf and custom scripts.
> 
> Example: Create a new vSwitch and assign it to a distro manually.
> 
> ⚠️ This is non-trivial and not officially supported.
> 
> Run one instance under WSL2 and another under WSL1
> 
> WSL1 instances have their own unique networking model (they use Windows’ network stack directly).
> 
> This way you can run 22.04 on WSL2 and 24.04 on WSL1 with distinct IP behavior.
> 
> Use Docker or VM approach for strict IP separation
> 
> If you truly need per-distro unique IPs (like for complex multi-host setups), it’s cleaner to use Docker networks or full Hyper-V VMs.
> 
> 📌 Bottom Line
> 
> By design, WSL2 distros share the same virtualized network interface → same IP.
> 
> For most workflows, unique ports per instance is the recommended solution.
> 
> If you absolutely need unique IPs, you’ll have to rely on advanced networking (custom Hyper-V NAT) or switch one distro to WSL1.
> 
> ✅ In short: WSL2 isn’t designed to give each distro its own IP — they all share the same VM’s network namespace. The practical solution is to differentiate services by ports or use WSL1/VMs for strict per-distro IP isolation.

What an AI grifter reply. There is no practical reason especially, no performance reason. 

VMs, are slow because they do full virtualization of everything including the kernel. Even then, in 2026 proper VMs are not slow.

In any case you dont need a full VM to get network isolation, as you noted its bound to network namespaces.

For sure, VMs are not slow because of an extra network namespace. You can try it yourself, `ip netns add foo` and see how quick this goes. 

Maybe there are some legit reasons, but its none of what you, or likely your LLM, said. I mean common look at these emojis.
